service/dap/server.go

```diff
@@ -69,6 +71,9 @@ type Server struct {
 	variableHandles *variablesHandlesMap
 	// args tracks special settings for handling debug session requests.
 	args launchAttachArgs
+	// exceptions maps goroutineIds to info about the last exception that occured.
+	// Reset at every stop.
+	exceptions map[int]dap.ExceptionInfoResponseBody
```

Comment:
IMHO it would be better to recover the informations about the exception inside `onExceptionInfoRequest` instead of having to keep another piece of global state correct. You can add a helper method to `debugger.(*Debugger)` to call `proc.FindGoroutine` and then the `Thread` member field will give you the thread of the goroutine with info about the current breakpoint.

Reply:
Yeah I agree I think that would be better. Do you think that would also work with the runtime errors?

I'll experiment a bit and see if I can make that work.

Reply:
It should, there is nothing special about them. The only thing I'm not sure about is if the current goroutine id is 0, but in that case you can just use CurrentThread.

Reply:
Thanks for the pointers!

I'm planning to address the issue with current goroutine id is 0 in a separate PR.  We are considering sending the current thread in addition to the goroutines in the threadsResponse when no goroutine is selected, so the exception info will correspond to a thread.

Reply:
Ok I was able to successfully move the handling of the two breakpoints to `onExceptionInfoRequest`, but the runtime errors as far as I can tell are not saved anywhere. We could either cache them on the debugger side (by saving them and then putting the result in the DebuggerState), or on the dap server side. Do you have any thoughts on which is better? Or of other ways to implement it that I may be missing?

Reply:
No, that error isn't saved anywhere. It's weird to associate it with a specific thread (or goroutine), because usually it isn't. But the one exception, bad access, is -- and it's probably the only error that's ever going to happen in that branch. I'd say save it on DAP's side and return it in onExceptionInfoRequest but only for the current thread/goroutine.